File: backend/app/utils/validation.py
"""
Input validation utilities for backend API endpoints
"""
import re
import mimetypes
from typing import List, Optional, Dict, Any, Union
from pathlib import Path
from fastapi import UploadFile, HTTPException, status
from pydantic import BaseModel, validator, ValidationError


class ValidationResult(BaseModel):
    """Result of validation operation"""
    is_valid: bool
    errors: List[str] = []
    warnings: List[str] = []


class FileValidationConfig(BaseModel):
    """Configuration for file validation"""
    max_size: int = 50 * 1024 * 1024  # 50MB
    min_size: int = 1024  # 1KB
    allowed_mime_types: List[str] = [
        'application/pdf',
        'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        'text/plain'
    ]
    allowed_extensions: List[str] = ['.pdf', '.docx', '.txt']
    max_filename_length: int = 255
    forbidden_patterns: List[str] = [r'\.\.', r'[<>:"|?*]', r'^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])$']


class TextValidationConfig(BaseModel):
    """Configuration for text validation"""
    min_length: int = 0
    max_length: int = 10000
    required: bool = False
    pattern: Optional[str] = None
    pattern_message: str = "Invalid format"


def validate_file(file: UploadFile, config: Optional[FileValidationConfig] = None) -> ValidationResult:
    """
    Validate uploaded file against configuration
    
    Args:
        file: FastAPI UploadFile object
        config: Validation configuration
        
    Returns:
        ValidationResult with validation status and errors
    """
    if config is None:
        config = FileValidationConfig()
    
    errors = []
    warnings = []
    
    # Check if file exists
    if not file or not file.filename:
        errors.append("No file provided")
        return ValidationResult(is_valid=False, errors=errors)
    
    # Validate filename
    filename = file.filename.strip()
    if not filename:
        errors.append("Filename cannot be empty")
    
    if len(filename) > config.max_filename_length:
        errors.append(f"Filename too long (max {config.max_filename_length} characters)")
    
    # Check for forbidden patterns in filename
    for pattern in config.forbidden_patterns:
        if re.search(pattern, filename, re.IGNORECASE):
            errors.append(f"Filename contains forbidden pattern: {pattern}")
    
    # Validate file extension
    file_extension = Path(filename).suffix.lower()
    if file_extension not in [ext.lower() for ext in config.allowed_extensions]:
        errors.append(
            f"File extension '{file_extension}' not allowed. "
            f"Allowed extensions: {', '.join(config.allowed_extensions)}"
        )
    
    # Validate MIME type
    content_type = file.content_type
    if content_type not in config.allowed_mime_types:
        # Try to guess MIME type from filename
        guessed_type, _ = mimetypes.guess_type(filename)
        if guessed_type and guessed_type in config.allowed_mime_types:
            warnings.append(f"MIME type mismatch, but filename suggests valid type: {guessed_type}")
        else:
            errors.append(
                f"File type '{content_type}' not allowed. "
                f"Allowed types: {', '.join(config.allowed_mime_types)}"
            )
    
    # Validate file size (if available)
    if hasattr(file, 'size') and file.size is not None:
        if file.size > config.max_size:
            errors.append(
                f"File size ({format_file_size(file.size)}) exceeds maximum "
                f"allowed size ({format_file_size(config.max_size)})"
            )
        
        if file.size < config.min_size:
            errors.append(
                f"File size ({format_file_size(file.size)}) below minimum "
                f"required size ({format_file_size(config.min_size)})"
            )
    
    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors,
        warnings=warnings
    )


def validate_text(text: str, config: Optional[TextValidationConfig] = None) -> ValidationResult:
    """
    Validate text input against configuration
    
    Args:
        text: Text to validate
        config: Validation configuration
        
    Returns:
        ValidationResult with validation status and errors
    """
    if config is None:
        config = TextValidationConfig()
    
    errors = []
    warnings = []
    
    # Handle None or non-string input
    if text is None:
        text = ""
    else:
        text = str(text)
    
    trimmed_text = text.strip()
    
    # Required field validation
    if config.required and not trimmed_text:
        errors.append("This field is required")
        return ValidationResult(is_valid=False, errors=errors)
    
    # Length validation (only if text is provided)
    if trimmed_text:
        if len(trimmed_text) < config.min_length:
            errors.append(f"Minimum length is {config.min_length} characters")
        
        if len(trimmed_text) > config.max_length:
            errors.append(f"Maximum length is {config.max_length} characters")
        
        # Pattern validation
        if config.pattern:
            try:
                if not re.match(config.pattern, trimmed_text):
                    errors.append(config.pattern_message)
            except re.error as e:
                errors.append(f"Invalid validation pattern: {str(e)}")
    
    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors,
        warnings=warnings
    )


def validate_email(email: str) -> ValidationResult:
    """Validate email address"""
    email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    config = TextValidationConfig(
        required=True,
        pattern=email_pattern,
        pattern_message="Please enter a valid email address"
    )
    return validate_text(email, config)


def validate_url(url: str, required: bool = False) -> ValidationResult:
    """Validate URL"""
    if not required and not url.strip():
        return ValidationResult(is_valid=True, errors=[])
    
    url_pattern = r'^https?:\/\/(?:[-\w.])+(?:\:[0-9]+)?(?:\/(?:[\w\/_.])*(?:\?(?:[\w&=%.])*)?(?:\#(?:[\w.])*)?)?$'
    config = TextValidationConfig(
        required=required,
        pattern=url_pattern,
        pattern_message="Please enter a valid URL (http:// or https://)"
    )
    return validate_text(url, config)


def validate_path(path: str) -> ValidationResult:
    """Validate file system path"""
    errors = []
    
    if not path.strip():
        errors.append("Path is required")
        return ValidationResult(is_valid=False, errors=errors)
    
    # Security checks
    if '..' in path:
        errors.append("Path cannot contain '..' for security reasons")
    
    if len(path) > 500:
        errors.append("Path is too long (maximum 500 characters)")
    
    # Check for invalid characters
    invalid_chars = ['<', '>', '"', '|', '?', '*']
    for char in invalid_chars:
        if char in path:
            errors.append(f"Path contains invalid character: '{char}'")
    
    # Check for reserved names (Windows)
    reserved_names = ['CON', 'PRN', 'AUX', 'NUL'] + [f'COM{i}' for i in range(1, 10)] + [f'LPT{i}' for i in range(1, 10)]
    path_parts = path.replace('\\', '/').split('/')
    for part in path_parts:
        if part.upper() in reserved_names:
            errors.append(f"Path contains reserved name: '{part}'")
    
    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors
    )


def validate_schema_type(schema_type: str) -> ValidationResult:
    """Validate schema type"""
    valid_schema_types = ['EU_ESRS_CSRD', 'UK_SRD', 'OTHER']
    
    if not schema_type:
        return ValidationResult(is_valid=False, errors=["Schema type is required"])
    
    if schema_type not in valid_schema_types:
        return ValidationResult(
            is_valid=False,
            errors=[f"Invalid schema type. Must be one of: {', '.join(valid_schema_types)}"]
        )
    
    return ValidationResult(is_valid=True, errors=[])


def validate_query(query: str) -> ValidationResult:
    """Validate search/RAG query"""
    config = TextValidationConfig(
        required=True,
        min_length=3,
        max_length=1000
    )
    return validate_text(query, config)


def validate_json_data(data: Dict[str, Any], required_fields: List[str] = None) -> ValidationResult:
    """Validate JSON data structure"""
    errors = []
    
    if not isinstance(data, dict):
        errors.append("Data must be a JSON object")
        return ValidationResult(is_valid=False, errors=errors)
    
    if required_fields:
        for field in required_fields:
            if field not in data:
                errors.append(f"Required field '{field}' is missing")
            elif data[field] is None or (isinstance(data[field], str) and not data[field].strip()):
                errors.append(f"Required field '{field}' cannot be empty")
    
    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors
    )


def validate_pagination_params(page: int, size: int, max_size: int = 100) -> ValidationResult:
    """Validate pagination parameters"""
    errors = []
    
    if page < 0:
        errors.append("Page number must be non-negative")
    
    if size <= 0:
        errors.append("Page size must be positive")
    
    if size > max_size:
        errors.append(f"Page size cannot exceed {max_size}")
    
    return ValidationResult(
        is_valid=len(errors) == 0,
        errors=errors
    )


def format_file_size(bytes_size: int) -> str:
    """Format file size in human-readable format"""
    if bytes_size == 0:
        return "0 Bytes"
    
    size_names = ["Bytes", "KB", "MB", "GB", "TB"]
    i = 0
    while bytes_size >= 1024 and i < len(size_names) - 1:
        bytes_size /= 1024.0
        i += 1
    
    return f"{bytes_size:.2f} {size_names[i]}"


def create_validation_error(validation_result: ValidationResult, field_name: str = None) -> HTTPException:
    """Create HTTPException from validation result"""
    message = "Validation failed"
    if field_name:
        message = f"Validation failed for field '{field_name}'"
    
    return HTTPException(
        status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
        detail={
            "message": message,
            "errors": validation_result.errors,
            "warnings": validation_result.warnings if validation_result.warnings else None
        }
    )


class ValidatedModel(BaseModel):
    """Base model with enhanced validation"""
    
    class Config:
        validate_assignment = True
        extra = "forbid"  # Forbid extra fields
    
    def validate_and_raise(self):
        """Validate model and raise HTTPException if invalid"""
        try:
            self.dict()  # This triggers validation
        except ValidationError as e:
            raise HTTPException(
                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
                detail={
                    "message": "Model validation failed",
                    "errors": [error["msg"] for error in e.errors()],
                    "details": e.errors()
                }
            )


# Decorator for endpoint validation
def validate_request(validator_func):
    """Decorator to validate request data"""
    def decorator(func):
        async def wrapper(*args, **kwargs):
            # Extract request data from kwargs
            request_data = {}
            for key, value in kwargs.items():
                if not key.startswith('_'):  # Skip internal parameters
                    request_data[key] = value
            
            # Validate request data
            validation_result = validator_func(request_data)
            if not validation_result.is_valid:
                raise create_validation_error(validation_result)
            
            return await func(*args, **kwargs)
        return wrapper
    return decorator

File: frontend/src/utils/validation.ts
// Input validation utilities

export interface ValidationResult {
  isValid: boolean;
  errors: string[];
}

export interface FileValidationOptions {
  maxSize?: number; // in bytes
  allowedTypes?: string[];
  allowedExtensions?: string[];
  minSize?: number;
}

// File validation
export const validateFile = (file: File, options: FileValidationOptions = {}): ValidationResult => {
  const errors: string[] = [];
  
  const {
    maxSize = 50 * 1024 * 1024, // 50MB default
    minSize = 1024, // 1KB minimum
    allowedTypes = ['application/pdf', 'application/vnd.openxmlformats-officedocument.wordprocessingml.document', 'text/plain'],
    allowedExtensions = ['.pdf', '.docx', '.txt']
  } = options;
  
  // Check file size
  if (file.size > maxSize) {
    errors.push(`File size (${formatFileSize(file.size)}) exceeds maximum allowed size (${formatFileSize(maxSize)})`);
  }
  
  if (file.size < minSize) {
    errors.push(`File size (${formatFileSize(file.size)}) is below minimum required size (${formatFileSize(minSize)})`);
  }
  
  // Check file type
  if (!allowedTypes.includes(file.type)) {
    errors.push(`File type "${file.type}" is not supported. Allowed types: ${allowedTypes.join(', ')}`);
  }
  
  // Check file extension
  const fileExtension = '.' + file.name.split('.').pop()?.toLowerCase();
  if (!allowedExtensions.some(ext => ext.toLowerCase() === fileExtension)) {
    errors.push(`File extension "${fileExtension}" is not supported. Allowed extensions: ${allowedExtensions.join(', ')}`);
  }
  
  // Check for empty file name
  if (!file.name.trim()) {
    errors.push('File name cannot be empty');
  }
  
  // Check for suspicious file names
  if (file.name.includes('..') || file.name.includes('/') || file.name.includes('\\')) {
    errors.push('File name contains invalid characters');
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
};

// Text input validation
export const validateText = (text: string, options: {
  minLength?: number;
  maxLength?: number;
  required?: boolean;
  pattern?: RegExp;
  patternMessage?: string;
} = {}): ValidationResult => {
  const errors: string[] = [];
  const {
    minLength = 0,
    maxLength = 10000,
    required = false,
    pattern,
    patternMessage = 'Invalid format'
  } = options;
  
  const trimmedText = text.trim();
  
  if (required && !trimmedText) {
    errors.push('This field is required');
  }
  
  if (trimmedText && trimmedText.length < minLength) {
    errors.push(`Minimum length is ${minLength} characters`);
  }
  
  if (trimmedText.length > maxLength) {
    errors.push(`Maximum length is ${maxLength} characters`);
  }
  
  if (pattern && trimmedText && !pattern.test(trimmedText)) {
    errors.push(patternMessage);
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
};

// Email validation
export const validateEmail = (email: string): ValidationResult => {
  const emailPattern = /^[a-zA-Z0-9][a-zA-Z0-9._%+-]*[a-zA-Z0-9]@[a-zA-Z0-9][a-zA-Z0-9.-]*[a-zA-Z0-9]\.[a-zA-Z]{2,}$/;
  return validateText(email, {
    required: true,
    pattern: emailPattern,
    patternMessage: 'Please enter a valid email address'
  });
};

// URL validation
export const validateUrl = (url: string, required = false): ValidationResult => {
  if (!required && !url.trim()) {
    return { isValid: true, errors: [] };
  }
  
  try {
    const urlObj = new URL(url);
    // Only allow http and https protocols
    if (!['http:', 'https:'].includes(urlObj.protocol)) {
      return { isValid: false, errors: ['Please enter a valid URL (http:// or https://)'] };
    }
    return { isValid: true, errors: [] };
  } catch {
    return { isValid: false, errors: ['Please enter a valid URL'] };
  }
};

// Path validation for remote directories
export const validatePath = (path: string): ValidationResult => {
  const errors: string[] = [];
  
  if (!path.trim()) {
    errors.push('Path is required');
    return { isValid: false, errors };
  }
  
  // Basic path validation
  if (path.includes('..')) {
    errors.push('Path cannot contain ".." for security reasons');
  }
  
  if (path.length > 500) {
    errors.push('Path is too long (maximum 500 characters)');
  }
  
  // Check for invalid characters (basic check)
  const invalidChars = ['<', '>', '"', '|', '?', '*'];
  const hasInvalidChars = invalidChars.some(char => path.includes(char));
  if (hasInvalidChars) {
    errors.push(`Path contains invalid characters: ${invalidChars.join(', ')}`);
  }
  
  return {
    isValid: errors.length === 0,
    errors
  };
};

// Schema type validation
export const validateSchemaType = (schemaType: string): ValidationResult => {
  const validSchemaTypes = ['EU_ESRS_CSRD', 'UK_SRD', 'OTHER'];
  
  if (!schemaType) {
    return { isValid: false, errors: ['Schema type is required'] };
  }
  
  if (!validSchemaTypes.includes(schemaType)) {
    return { 
      isValid: false, 
      errors: [`Invalid schema type. Must be one of: ${validSchemaTypes.join(', ')}`] 
    };
  }
  
  return { isValid: true, errors: [] };
};

// Query validation for search and RAG
export const validateQuery = (query: string): ValidationResult => {
  return validateText(query, {
    required: true,
    minLength: 3,
    maxLength: 1000,
  });
};

// Utility function to format file size
export const formatFileSize = (bytes: number): string => {
  if (bytes === 0) return '0 Bytes';
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  const formattedSize = (bytes / Math.pow(k, i));
  
  // Only show decimals if they're not zero
  if (formattedSize % 1 === 0) {
    return formattedSize + ' ' + sizes[i];
  } else {
    return formattedSize.toFixed(2) + ' ' + sizes[i];
  }
};

import React from 'react';

// Real-time validation hook
export const useRealTimeValidation = (
  value: string,
  validator: (value: string) => ValidationResult,
  debounceMs = 300
) => {
  const [validationResult, setValidationResult] = React.useState<ValidationResult>({ isValid: true, errors: [] });
  const [isValidating, setIsValidating] = React.useState(false);
  
  React.useEffect(() => {
    setIsValidating(true);
    const timer = setTimeout(() => {
      const result = validator(value);
      setValidationResult(result);
      setIsValidating(false);
    }, debounceMs);
    
    return () => clearTimeout(timer);
  }, [value, validator, debounceMs]);
  
  return { validationResult, isValidating };
};

// Form validation helper
export const validateForm = (fields: Record<string, any>, validators: Record<string, (value: any) => ValidationResult>): {
  isValid: boolean;
  errors: Record<string, string[]>;
} => {
  const errors: Record<string, string[]> = {};
  let isValid = true;
  
  Object.keys(validators).forEach(fieldName => {
    const validator = validators[fieldName];
    const value = fields[fieldName];
    const result = validator(value);
    
    if (!result.isValid) {
      errors[fieldName] = result.errors;
      isValid = false;
    }
  });
  
  return { isValid, errors };
};